### PR TITLE
[SRVKS-944] Inject net-kourier secret informer filtering env var

### PR DESCRIPTION
**Pull request by skonto**
## Proposed Changes

- Similar to https://github.com/openshift-knative/serverless-operator/pull/1616 but using env vars instead
- Net kourier patch is available in 1.5: https://github.com/openshift-knative/net-kourier/pull/32

### CODE REVIEW

**Comment by nak3** [comment]
I assumed that we would enable this config by default when you implemented. What do you think that enabling it by default instead of adding this UX improvement?

**Comment by skonto** [comment]
> What do you think that enabling it by default instead of adding this UX improvement?

We can try. I am not sure to what extent customers have tested the net-istio fix.

**Comment by skonto** [comment]
@nak3 I added the following:
- enable by default if no annotation is present
- do validation of the annotation at the CRD side
As I mentioned above the idea is to deprecate and remove the annotation in a future release assuming things are ok at the customer side. I want to have an option to disable it though.

**Comment by skonto** [comment]
Pls review, /assign @nak3, @mgencur

**Comment by skonto** [comment]
Need to fix the tests by adding to secrets the right annotation.

File: test/upgrade/installation/serverless.go [hunk 1]
```diff
@@ -167,7 +167,7 @@ func moveCRDsToAlpha(ctx *test.Context, name string) error {
 	crd.Spec.Conversion = &apiextension.CustomResourceConversion{Strategy: apiextension.ConversionStrategyType("None")}
 	_, err = ctx.Clients.APIExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Update(context.Background(), crd, metav1.UpdateOptions{})
 	if err != nil {
-		return err
+		return fmt.Errorf("moving crd to alpha failed: %w", err)
```
Comment (skonto) [inline test/upgrade/installation/serverless.go:170]: Added to debug the downgrade test failure.

**Comment by skonto** [comment]
Debugging ```{Failed  === RUN   TestServerlessDowngrade/DowngradeWith/DowngradeServerless
    upgrade_test.go:194: Serverless downgrade failed: Operation cannot be fulfilled on customresourcedefinitions.apiextensions.k8s.io "knativeservings.operator.knative.dev": the object has been modified; please apply your changes to the latest version and try again
        --- FAIL: TestServerlessDowngrade/DowngradeWith/DowngradeServerless (0.65s)```
Once everything is fixed manifest changes and patching will be moved to the midstream repo.

**Comment by skonto** [comment]
`TestServerlessUpgrade/VerifyContinualTests/KafkaSinkSourceContinualTest`

**Comment by skonto** [comment]
Once https://github.com/openshift/knative-serving/pull/1281 is merged we will remove parts of it. We will not update openshift-knative/net-{kourier, repos) for now to set this enabled by default. This will happen later on.

File: test/serving.bash [hunk 2]
```diff
@@ -30,6 +30,11 @@ function prepare_knative_serving_tests {
 
   export GATEWAY_OVERRIDE="kourier"
   export GATEWAY_NAMESPACE_OVERRIDE="${INGRESS_NAMESPACE}"
+
+  # TODO: move to midstream when we enable by default there
+  sed -i '/Name: test.AppendRandomString("byocert-secret"),/a Labels: map[string]string{ "networking.internal.knative.dev/certificate-uid": "byocert-secret", },' \
```
Comment (skonto) [inline test/serving.bash:35]: No need for this after midstream PR is merged.

File: openshift-knative-operator/pkg/serving/istio.go [hunk 3]
```diff
@@ -7,10 +7,23 @@ import (
 	"knative.dev/operator/pkg/apis/operator/base"
```
Comment (skonto) [inline openshift-knative-operator/pkg/serving/istio.go:7]: File will probably needs to be renamed as we cover both net-kourier and net-istio.

**Comment by nak3** [comment]
When users added `ENABLE_SECRET_INFORMER_FILTERING_BY_CERT_UID` via `env` in `KnativeServing` CR, it will not be conflicted?

I am still thinking we just should use `common.ConfigureIfUnset` to configure it via env rather than keep supporting `"serverless.openshift.io/enable-secret-informer-filtering"`.

**Comment by skonto** [comment]
```
{"severity":"ERROR","timestamp":"2022-11-03T11:59:28.801435925Z","logger":"net-istio-controller.istio-ingress-controller","caller":"ingress/reconciler.go:313","message":"Returned an error","knative.dev/controller":"istio-ingress-controller","knative.dev/controller":"knative.dev.net-istio.pkg.reconciler.ingress.Reconciler","knative.dev/kind":"networking.internal.knative.dev.Ingress","knative.dev/traceid":"baa25261-95c3-4149-89ea-6593e6fee811","knative.dev/key":"serving-tests/b-y-o-certificate-zcowmplt.apps84215f9df0db30751afef2a297f08571","targetMethod":"ReconcileKind","error":"secret \"byocert-secret-cjegwyws\" not found","stacktrace":"knative.dev/networking/pkg/client/injection/reconciler/networking/v1alpha1/ingress.(*reconcilerImpl).Reconcile\n\t/app/vendor/knative.dev/networking/pkg/client/injection/reconciler/networking/v1alpha1/ingress/reconciler.go:313\nknative.dev/pkg/controller.
```

**Comment by skonto** [comment]
https://github.com/openshift/knative-serving/pull/1281 is in so we dont need the patch any more.

File: openshift-knative-operator/pkg/serving/secretinformerfiltering.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+			}
+			labels[networking.CertificateUIDLabelKey] = "data-plane"
+			u.SetLabels(labels)
+			return nil
+		}
+		return nil
+	}
+}
+
+// Adds default (true) to env vars for secret informer filtering in net-* deployments and returns if we should inject
```
Comment (nak3) [inline openshift-knative-operator/pkg/serving/secretinformerfiltering.go:49]: Sorry if I misunderstood but we will add the default to `true` on v1.26?
I thought we will move it to 1.27 as per your comment in https://issues.redhat.com/browse/SRVKS-944

Comment (skonto) [reply openshift-knative-operator/pkg/serving/secretinformerfiltering.go:49]: I think docs team still has some time. Let me pinged them again. Doc issue is blocked on eng at the moment.

**Comment by skonto** [comment]
@nak3 gentle ping for a stamp.

**Comment by nak3** [comment]
/lgtm

:+1:

**Pull request merged**